src/symmetry.jl

```diff
@@ -0,0 +1,220 @@
+## This file contains functions to handle the symetries
+## See the docs for the theoretical explanation
+
+# A symmetry operation (symop) is a couple (Stilde, τtilde) of a
+# unitary (in cartesian coordinates, but not in reduced coordinates)
+# matrix Stilde and a translation τtilde such that, for each atom of
+# type A at position a, Stilde a + τtilde is also an atom of type A.
+
+# This induces a symmetry in the Brillouin zone that the Hamiltonian
+# at S k is unitary equivalent to that at k, which we exploit to
+# reduce computations. The relationship is
+# S = Stilde'
+# τ = -Stilde^-1 τtilde
+# (valid both in reduced and cartesian coordinates)
+
+# There is by decreasing cardinality
+# - The group of symmetry operations of the lattice
+# - The group of symmetry operations of the crystal
+# - The group of symmetry operations of the crystal that preserves the BZ mesh
+# - The set of symmetry operations that we use to reduce the RBZ to the IBZ
+
+# (S,τ)
+const SymOp = Tuple{Mat3{Int}, Vec3{Float64}}
+
+@doc raw"""
+Return the ``k``-point symmetry operations associated to a lattice, model or basis.
+Since the ``k``-point discretisations may break some of the symmetries, the latter
+case will return a subset of the symmetries of the former two.
+"""
+function symmetry_operations(lattice, atoms; tol_symmetry=1e-5, kcoords=nothing)
+    symops = []
+    # Get symmetries from spglib
+    Stildes, τtildes = spglib_get_symmetry(lattice, atoms, tol_symmetry=tol_symmetry)
+
+    # Notice: In the language of the latex document in the docs
+    # spglib returns \tilde{S} and \tilde{τ} in integer real-space coordinates, such that
+    # (A Stilde A^{-1}) is the actual \tilde{S} from the document as a unitary matrix.
+    #
+    # Still we have the following properties for S and τ given in *integer* and
+    # *fractional* real-space coordinates:
+    #      \tilde{S}^{-1} = S^T (if applied to a vector in frac. coords in reciprocal space)
+
+    for isym = 1:length(Stildes)
+        S = Stildes[isym]'                  # in fractional reciprocal coordinates
+        τ = -Stildes[isym] \ τtildes[isym]  # in fractional real-space coordinates
+        τ = τ .- floor.(τ)
+        @assert all(0 .≤ τ .< 1)
+        push!(symops, (S, τ))
+    end
+
+    symops = unique(symops)
+
+    if kcoords !== nothing
+        # filter only the operations that respect the symmetries of the discrete BZ grid
+        function preserves_grid(S)
+            all(normalize_kpoint_coordinate(S * k) in kcoords
```

Comment:
Probably normalize_kpoint_coordinate should live in this file as well?

Reply:
Not really, normalize has nothing to do with symmetries. It's fine that symmetry.jl and bzmesh.jl depend from each other I think

Reply:
hmm ... I prefer hierarchical dependencies. Maybe normalize_kpoint_coordinate should go to a third file?

Reply:
I thought the exact same thing, but after thinking about it I'm pretty sure it's leftover from static languages. There's very little dependency problems in julia, you just need to define datastructures before functions (eg the symops declaration will have to go to types.jl if we want to use it in Model) but for functions there's no such constraints. The real question is: will somebody reading this be able to figure out where this function is defined? I think bzmesh.jl is a pretty natural place to look for.

Reply:
You might have a point there. In either way our code is still small enough that it's not too hard to grep for a function. Let's leave it like that for now.

Reply:
Btw it's a point against the index: it screws up my `grep -ir` at DFTK toplevel :-p